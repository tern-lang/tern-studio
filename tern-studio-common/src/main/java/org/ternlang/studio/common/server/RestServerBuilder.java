package org.ternlang.studio.common.server;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.net.InetSocketAddress;
import java.util.List;
import java.util.Optional;

import javax.net.ssl.SSLContext;
import javax.ws.rs.ProcessingException;
import javax.ws.rs.core.Application;

import org.glassfish.jersey.server.ResourceConfig;
import org.glassfish.jersey.simple.SimpleContainer;
import org.glassfish.jersey.simple.SimpleServer;
import org.glassfish.jersey.simple.SimpleTraceAnalyzer;
import org.glassfish.jersey.simple.internal.LocalizationMessages;
import org.simpleframework.http.core.Container;
import org.simpleframework.http.core.ContainerSocketProcessor;
import org.simpleframework.http.socket.service.Router;
import org.simpleframework.http.socket.service.RouterContainer;
import org.simpleframework.http.socket.service.Service;
import org.simpleframework.transport.SocketProcessor;
import org.simpleframework.transport.connect.Connection;
import org.simpleframework.transport.connect.SocketConnection;
import org.springframework.stereotype.Component;
import org.ternlang.studio.common.ProgressManager;
import org.ternlang.studio.common.resource.CombinationMatcher;
import org.ternlang.studio.common.resource.ResourcePathRouter;

@Component
public class RestServerBuilder {
   
   private final CombinationMatcher matcher;
   private final Router router;

   public RestServerBuilder(CombinationMatcher matcher, Optional<List<Service>> services) {
      this.router = new ResourcePathRouter(services);
      this.matcher = matcher;
   }
   
   public SimpleServer create(ResourceConfig config, SSLContext context, int port) throws Exception {
      SimpleContainer container = create(config);
      Container inner = create(container);
      
      try {
         SimpleTraceAnalyzer analyzer = new SimpleTraceAnalyzer();
         SocketProcessor server = new ContainerSocketProcessor(inner);
         Connection connection = new SocketConnection(server, analyzer);
         InetSocketAddress listen = new InetSocketAddress(port);
         InetSocketAddress bound = (InetSocketAddress)connection.connect(listen, context);
         int bindPort = bound.getPort();
         
         container.getApplicationHandler().onStartup(container);
         ProgressManager.getProgress().update("Container started on " + bindPort);

         return new InternalServer(container, analyzer, connection, bindPort);
      } catch (final IOException ex) {
         throw new ProcessingException(LocalizationMessages.ERROR_WHEN_CREATING_SERVER(), ex);
      }
   }
   
   private Container create(SimpleContainer container) {
      try {
         RestServerContainer delegate = new RestServerContainer(container, matcher);
   
         if(router != null) {
            return new RouterContainer(delegate, router, 5);
         }
         return delegate;
      } catch(Exception e) {
         throw new IllegalStateException("Could not create router container", e);
      }
   }
   
   private SimpleContainer create(ResourceConfig config) {
      try {
         Constructor<SimpleContainer> constructor = SimpleContainer.class.getDeclaredConstructor(Application.class);
         constructor.setAccessible(true);
         return constructor.newInstance(config);
      } catch(Exception e) {
         throw new IllegalStateException("Could not create container", e);
      }
   }
   
   private static class InternalServer implements SimpleServer {
      
      private final SimpleTraceAnalyzer analyzer;
      private final SimpleContainer container;
      private final Connection connection;
      private final int port;
      
      public InternalServer(SimpleContainer container, SimpleTraceAnalyzer analyzer, Connection connection, int port) {
         this.analyzer = analyzer;
         this.container = container;
         this.connection = connection;
         this.port = port;
      }
      
      @Override
      public void close() throws IOException {
         container.getApplicationHandler().onShutdown(container);
         analyzer.stop();
         connection.close();
      }

      @Override
      public int getPort() {
         return port;
      }

      @Override
      public boolean isDebug() {
         return analyzer.isActive();
      }

      @Override
      public void setDebug(boolean enable) {
         if (enable) {
            analyzer.start();
         } else {
            analyzer.stop();
         }
      }
      
   }
}
