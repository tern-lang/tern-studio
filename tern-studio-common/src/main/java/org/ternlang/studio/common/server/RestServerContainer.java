package org.ternlang.studio.common.server;

import static org.simpleframework.http.Method.CONNECT;
import static org.simpleframework.http.Protocol.DATE;
import static org.simpleframework.http.Protocol.SERVER;
import static org.ternlang.studio.common.resource.SessionConstants.SESSION_ID;

import java.io.IOException;
import java.util.UUID;

import lombok.extern.slf4j.Slf4j;

import org.glassfish.jersey.simple.SimpleContainer;
import org.simpleframework.http.Cookie;
import org.simpleframework.http.Request;
import org.simpleframework.http.Response;
import org.simpleframework.http.Status;
import org.simpleframework.http.core.Container;
import org.ternlang.studio.common.resource.CombinationMatcher;
import org.ternlang.studio.common.resource.Resource;

@Slf4j
public class RestServerContainer implements Container {
   
   private static final String SERVER_NAME = "Apache/2.2.14";

   private final CombinationMatcher matcher;
   private final SimpleContainer container;

   public RestServerContainer(SimpleContainer container, CombinationMatcher matcher) {
      this.matcher = matcher;
      this.container = container;
   }

   @Override
   public void handle(Request request, Response response) {
      long time = System.currentTimeMillis();
      String method = request.getMethod();
      
      try {
         Resource resource = matcher.match(request, response);
         Cookie cookie = request.getCookie(SESSION_ID);
         
         if(cookie == null) {
            String value = UUID.randomUUID().toString();
            response.setCookie(SESSION_ID, value);
         }
         response.setDate(DATE, time);
         response.setValue(SERVER, SERVER_NAME);
         response.setDate(DATE, time);
         
         if(resource != null) {
            response.setStatus(Status.OK);
            resource.handle(request, response);
         } else {
            container.handle(request, response);
         }
      } catch (Throwable cause) {
         log.info("Error handling resource", cause);
      } finally {
         try {
            if(!method.equals(CONNECT)) {
               response.close();
            }
         } catch (IOException ignore) {
            log.info("Could not close response", ignore);
         }
      }
   }
}