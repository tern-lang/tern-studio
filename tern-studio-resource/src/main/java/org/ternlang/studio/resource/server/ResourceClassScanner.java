package org.ternlang.studio.resource.server;

import static org.springframework.core.io.support.ResourcePatternResolver.CLASSPATH_ALL_URL_PREFIX;

import java.lang.annotation.Annotation;
import java.net.URL;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.stereotype.Component;
import org.ternlang.core.Any;
import org.ternlang.core.ContextClassLoader;

import lombok.extern.slf4j.Slf4j;

@Slf4j
@org.ternlang.studio.resource.action.annotation.Component
@Component
public class ResourceClassScanner {
   
   private static final String RESOURCE_PACKAGE = "org.ternlang.";
   private static final String RESOURCE_PATH = "org/ternlang/";
   
   private final PathMatchingResourcePatternResolver resolver;
   private final ClassLoader loader;
   
   public ResourceClassScanner() {
      this.resolver = new PathMatchingResourcePatternResolver(); 
      this.loader = new ContextClassLoader(Any.class);
   }

   public Set<Class> scan(Class<? extends Annotation> annotation) {
      return scan(annotation, "");
   }
   
   public Set<Class> scan(Class<? extends Annotation> annotation, String suffix) {
      try {
         String pattern = CLASSPATH_ALL_URL_PREFIX + RESOURCE_PATH + "**/*" + suffix + ".class";
         Resource[] resources = resolver.getResources(pattern);
         
         log.debug("Scan of '{}' found {} resources", pattern, resources.length);
         
         if(resources.length > 0) {
            Set<Class> matches = new HashSet<Class>();
            
            for(Resource resource : resources) {
               URL target = resource.getURL();
               String location = target.toString();
               String name = resolve(location); // remove .class and path prefix
               
               if(name.startsWith(RESOURCE_PACKAGE)) {
                  try {
                     Class<?> type = loader.loadClass(name);
          
                     if(type.isAnnotationPresent(annotation)) {                        
                        log.debug("Loading resource {}", resource);
                        matches.add(type);
                     } else {
                        log.debug("Ignoring resource {}", resource);
                     }
                  } catch(Exception e) {
                     log.warn("Could not load {}", name);
                  }
               }
            }
            return matches;
         }
      } catch(Exception e) {
         throw new IllegalStateException("Could not scan for resources", e);
      }
      return Collections.emptySet();
   }
   
   private String resolve(String location) {
      try {
         int separator = location.lastIndexOf("!"); // if its a jar resource
         
         if(separator != -1) {
            location = location.substring(separator + 1);
         }
         String normal = location.replace('/', '.');
         int index = normal.lastIndexOf(RESOURCE_PACKAGE);
         int length = normal.length();
         
         return normal.substring(index, length - 6); // remove .class and path prefix
      } catch(Exception e) {
         throw new IllegalStateException("Could parse " + location, e);
      }
   }
}
