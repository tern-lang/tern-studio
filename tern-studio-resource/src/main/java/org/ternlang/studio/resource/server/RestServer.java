package org.ternlang.studio.resource.server;

import java.net.InetSocketAddress;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@org.ternlang.studio.resource.action.annotation.Component
@Component
public class RestServer {
   
   private final RestServerBuilder builder;
   private final int port;
   
   public RestServer(RestServerBuilder builder, @org.ternlang.studio.resource.action.annotation.Value("${port}") @Value("${port}") int port) {
      this.builder = builder;
      this.port = port;
   }

   public InetSocketAddress start() {
      try {
         return builder.create(null, port);
      } catch(Exception e) {
         throw new IllegalStateException("Could not create container", e);
      }
   }
}
