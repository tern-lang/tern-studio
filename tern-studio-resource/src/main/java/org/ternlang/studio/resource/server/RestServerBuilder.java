package org.ternlang.studio.resource.server;

import java.net.InetSocketAddress;

import javax.net.ssl.SSLContext;

import org.simpleframework.http.core.ContainerSocketProcessor;
import org.simpleframework.http.socket.service.Router;
import org.simpleframework.http.socket.service.RouterContainer;
import org.simpleframework.transport.SocketProcessor;
import org.simpleframework.transport.connect.Connection;
import org.simpleframework.transport.connect.SocketConnection;
import org.springframework.stereotype.Component;
import org.ternlang.studio.resource.ResourceMatcher;
import org.ternlang.studio.resource.ResourcePathRouter;
import org.ternlang.studio.resource.ResourceSystem;

@Component
public class RestServerBuilder {
   
   private final ResourceMatcher matcher;
   private final Router router;

   public RestServerBuilder(ResourceSystem system, ResourcePathRouter router) {
      this.matcher = system.create();
      this.router = router;
   }
   
   public InetSocketAddress create(SSLContext context, int port) throws Exception {
      RestServerContainer container = new RestServerContainer(matcher);
      RouterContainer wrapper = new RouterContainer(container, router, 5);
      SocketProcessor server = new ContainerSocketProcessor(wrapper);
      Connection connection = new SocketConnection(server);
      InetSocketAddress listen = new InetSocketAddress(port);
      
      return (InetSocketAddress)connection.connect(listen, context);
   }
}
