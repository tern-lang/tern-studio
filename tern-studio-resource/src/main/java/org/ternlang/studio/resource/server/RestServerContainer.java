package org.ternlang.studio.resource.server;

import static org.simpleframework.http.Method.CONNECT;
import static org.simpleframework.http.Protocol.DATE;
import static org.simpleframework.http.Protocol.SERVER;
import static org.ternlang.studio.resource.SessionConstants.SESSION_ID;

import java.io.IOException;
import java.util.UUID;

import org.simpleframework.http.Cookie;
import org.simpleframework.http.Request;
import org.simpleframework.http.Response;
import org.simpleframework.http.Status;
import org.simpleframework.http.core.Container;
import org.ternlang.studio.resource.Resource;
import org.ternlang.studio.resource.ResourceMatcher;

import lombok.extern.slf4j.Slf4j;

@Slf4j
public class RestServerContainer implements Container {
   
   private static final String SERVER_NAME = "Apache/2.2.14";

   private final ResourceMatcher matcher;

   public RestServerContainer(ResourceMatcher matcher) {
      this.matcher = matcher;
   }

   @Override
   public void handle(Request request, Response response) {
      long time = System.currentTimeMillis();
      String method = request.getMethod();
      
      try {
         Resource resource = matcher.match(request, response);
         Cookie cookie = request.getCookie(SESSION_ID);
         
         if(cookie == null) {
            String value = UUID.randomUUID().toString();
            response.setCookie(SESSION_ID, value);
         }
         response.setDate(DATE, time);
         response.setValue(SERVER, SERVER_NAME);
         response.setDate(DATE, time);
         
         if(resource != null) {
            response.setStatus(Status.OK);
            resource.handle(request, response);
         } else {
            throw new RuntimeException("Default response here " + request);
         }
      } catch (Throwable cause) {
         log.info("Error handling resource", cause);
      } finally {
         try {
            if(!method.equalsIgnoreCase(CONNECT)) {
               response.close();
            }
         } catch (IOException ignore) {
            log.info("Could not close response", ignore);
         }
      }
   }
}