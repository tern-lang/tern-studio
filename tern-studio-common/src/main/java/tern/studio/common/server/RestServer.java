package tern.studio.common.server;

import java.util.Set;

import javax.inject.Provider;
import javax.ws.rs.core.Feature;
import javax.ws.rs.core.FeatureContext;

import org.glassfish.hk2.api.ServiceLocator;
import org.glassfish.jersey.ServiceLocatorProvider;
import org.glassfish.jersey.server.ResourceConfig;
import org.glassfish.jersey.simple.SimpleServer;
import org.jvnet.hk2.spring.bridge.api.SpringBridge;
import org.jvnet.hk2.spring.bridge.api.SpringIntoHK2Bridge;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Component;

@Component
public class RestServer {
   
   private final ResourceClassScanner scanner;
   private final RestServerBuilder builder;
   private final int port;
   
   public RestServer(RestServerBuilder builder, @Value("${port}") int port) {
      this.scanner = new ResourceClassScanner();
      this.builder = builder;
      this.port = port;
   }

   public SimpleServer start(ApplicationContext context) {
      Set<Class<?>> resources = scanner.scan();
      
      try {
         ResourceConfig config = new ResourceConfig(resources);
         RestServiceLocatorFeature feature = new RestServiceLocatorFeature((ConfigurableApplicationContext)context);
 
         config.register(feature);
         
         return builder.create(config, null, port);
      } catch(Exception e) {
         throw new IllegalStateException("Could not create container", e);
      }
   }
   
   private static class RestServiceLocatorFeature implements Feature, Provider<ServiceLocator> {

      private ConfigurableApplicationContext spring;
      private ServiceLocator locator;

      public RestServiceLocatorFeature(ConfigurableApplicationContext spring) {
         this.spring = spring;
      }

      @Override
      public ServiceLocator get() {
         if(locator == null) {
            throw new IllegalStateException("Service locator is not yet available");
         }
         return locator;
      }

      @Override
      public boolean configure(FeatureContext context) {
         locator = ServiceLocatorProvider.getServiceLocator(context);

         SpringBridge.getSpringBridge().initializeSpringBridge(locator);
         SpringIntoHK2Bridge springBridge = locator.getService(SpringIntoHK2Bridge.class);
         springBridge.bridgeSpringBeanFactory(spring.getBeanFactory());

         return true;
      }
   }
}
