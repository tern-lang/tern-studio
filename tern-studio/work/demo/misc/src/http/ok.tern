import org.simpleframework.http.Status;
import http.server.Server;


Server.create(8180).register("/.*", (request, response) -> {
   println(request);
   response.setStatus(Status.OK);
   response.setValue("Server", "OK/1.0");
   response.setDate("Date", System.currentTimeMillis());
   response.setContentType("text/plain");
   response.getPrintStream().println(request);
   response.close();
}).start();

sleep(100000000);
