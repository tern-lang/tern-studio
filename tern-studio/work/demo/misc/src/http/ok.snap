import org.simpleframework.http.Status;
import http.Server;


Server.create(8180).register("/.*", (request, response) -> {
   println(request);
   response.setStatus(Status.OK);
   response.setValue("Server", "OK/1.0");
   response.setDate("Date", System.currentTimeMillis());
   response.setContentType("text/plain");
   response.getPrintStream().println("ok");
   response.close();
}).start();

sleep(100000000);
